firstapp/app/src/androidTest/java/com/example/firstapp/ShellUtility.java

```diff
@@ -138,55 +158,123 @@ public static UiObject castToObject(UiObject2 object2) throws invalidInputExcept
+     */
+    public UiObject executeUncachedAction(Action action, int i) throws invalidInputException, UiObjectNotFoundException, IOException, InterruptedException {
+        boolean strict = action.strict;
+
+        UiObject2 object2 = null;
+        UiObject object = null;
+        if (action instanceof StartAction) {
+            String pkg = ((StartAction) action).pkg;
+            forceQuitApp(pkg);
+            launchApp(pkg);
+        } else if (action instanceof ClickAction) {
+            String text = ((ClickAction) action).text;
+            if (strict) {
+                object2 = device.wait(Until.findObject(By.text(text)), timeoutMs);
+            } else {
+                object2 = device.wait(Until.findObject(By.textContains(text)), timeoutMs);
+            }
+            object = castToObject(object2);
+            object.click();
+        } else if(action instanceof ClickImageAction) {
+            String description = ((ClickImageAction) action).description;
+            if (strict) {
+                object2 = device.wait(Until.findObject(By.desc(description)), timeoutMs);
+            } else {
+                object2 = device.wait(Until.findObject(By.descContains(description)), timeoutMs);
+            }
+            object = castToObject(object2);
+            object.click();
+        } else if(action instanceof EditAction) {
+            String text = ((EditAction) action).text;
+            String entered = ((EditAction) action).entered;
+            if (strict) {
+                object2 = device.wait(Until.findObject(By.text(text)), timeoutMs);
+            } else {
+                object2 = device.wait(Until.findObject(By.textContains(text)), timeoutMs);
+            }
+            object = getEditableObject(object2);
+            object.waitForExists(timeoutMs);
+            object.legacySetText(entered);
+        } else {
+            throw new invalidInputException("an action has an invalid first token at index" + i);
+        }
+        return object;
+    }
+
+    /**
+     * Executes CUJ and returns cached objects
+     */
+    public UiObject[] cacheCUJ(Action[] CUJ) throws InterruptedException, invalidInputException, UiObjectNotFoundException, IOException {
+        UiObject[] cachedObjects = new UiObject[CUJ.length];
+        for (int i = 0; i < CUJ.length; i++) {
+            cachedObjects[i] = executeUncachedAction(CUJ[i], i);
+        }
+        return cachedObjects;
+    }
+
+    /**
+     * Executes a CUJ with preparatory actions pre and measured actions post. Caches information for later use.
+     */
+    public void executeCUJ(String[] preCUJ, String[] postCUJ) throws Exception {
+        String[] CUJStrings = new String[preCUJ.length + postCUJ.length];
+        System.arraycopy(preCUJ, 0, CUJStrings, 0, preCUJ.length);
+        System.arraycopy(postCUJ, 0, CUJStrings, preCUJ.length, postCUJ.length);
+        Action[] CUJ = parseStringCUJ(CUJStrings);
```

Comment:
Variable names shouldn't be all capitals, because it can be confused with class names.  They should always start with a lowercase letter.